src/inverted_index.c

```diff
@@ -141,11 +141,123 @@ ENCODER(encodeDocIdsOnly) {
   return WriteVarint(delta, bw);
 }
 
+/**
+ * Numeric Encoding Design:
+ *
+ * Internally, the number is always normalized as a double, however it may
+ * be stored in different ways depending on the actual size of the number.
+ *
+ * The tax paid up-front for this flexibility is a single byte; however this
+ * byte is never actually wasted.
+ *
+ * The encoding information is stored using the last two bits of the leading
+ * byte, giving us a possibility of 4 different types of encoding.
+ *
+ * For integer values under 2^6, the first 6 bits are used
+ *
+ *
+ */
+typedef enum {
+  // The actual value is a small integer, found in the 6 LSB
+  // DocID is Varint
+  NumEncoding_Tiny = 0x00 << 6,
+
+  // The value is found as qint(docid,num).
+  // Signdedness is determined by the 3rd MSB bit in the header.
+  NumEncoding_Int32 = 0x01 << 6,
+
+  // The value is found as qint(docid,num).
+  // Signdedness is determined by the 3rd MSB in the header
+  NumEncoding_Float32 = 0x02 << 6,
+
+  // The value is found as qint32_64(docid, u64).
+  // If the number is signed, it is encoded as a double
+  NumEncoding_Extended = 0x03 << 6
+} NumEncodingType;
+
+#define NUM_TINY_MAX 0x3F            // Mask/Limit for 'Tiny' value
+#define NUM_F_SIGNED 0x20            // If encoding is not extended, signed bit
+#define NUM_ENCODING_MASK 0xC0       // Mask for two encoding bits
+#define NUM_QINT_VALUEHDR_MASK 0x3C  // Mask for qint value header
+
+typedef union {
+  float f32;
+  uint32_t encoded;
+} NumEncoded;
+
+static inline double myAbsf64(double input) {
+  return input < 0 ? -input : input;
+}
+
+static void dumpBits(uint64_t value, size_t numBits, FILE *fp) {
+  while (numBits) {
+    fprintf(fp, "%d", !!(value & (1 << (numBits - 1))));
+    numBits--;
+  }
+}
+
 // 9. Special encoder for numeric values
 ENCODER(encodeNumeric) {
-  size_t sz = WriteVarint(delta, bw);
+  const double absVal = myAbsf64(res->num.value);
+  const double realVal = res->num.value;
+
+  const uint32_t u32Num = (uint32_t)absVal;
+  const float f32Num = (float)absVal;
+  const uint64_t u64Num = (uint64_t)absVal;
+  const uint8_t u8Num = ((uint8_t)realVal) & NUM_TINY_MAX;
+  size_t sz = 0;
 
-  sz += Buffer_Write(bw, (char *)&res->num.encoded, sizeof(uint32_t));
+  if ((double)u8Num == realVal) {
+    uint8_t b = NumEncoding_Tiny | u8Num;
+    sz += Buffer_Write(bw, &b, 1);
+    sz += WriteVarint(delta, bw);
+    return sz;
+  }
+
+  uint8_t encoding = 0;
+  size_t pos = bw->buf->offset;
+
+  if ((double)u32Num == absVal) {
+    // -UINT32_MAX <= N <= UINT32_MAX
+    sz += qint_encode2(bw, delta, u32Num);
```

Comment:
In this case you are writing the delta to bw without saving an extra byte for the encoding, isn't it?